src/builtins.c

```diff
@@ -1374,19 +1396,34 @@ static size_t jl_static_show_x_(JL_STREAM *out, jl_value_t *v, jl_datatype_t *vt
         n += jl_printf(out, "\"%s\"", jl_iostr_data(v));
     }
     else if (vt == jl_uniontype_type) {
-        n += jl_show_svec(out, ((jl_uniontype_t*)v)->types, "Union", "{", "}");
-    }
-    else if (vt == jl_typector_type) {
-        n += jl_static_show_x(out, ((jl_typector_t*)v)->body, depth);
-    }
-    else if (vt == jl_tvar_type) {
-        if (((jl_tvar_t*)v)->lb != jl_bottom_type) {
-            n += jl_static_show(out, ((jl_tvar_t*)v)->lb);
+        n += jl_printf(out, "Union{");
+        while (jl_is_uniontype(v)) {
+            n += jl_static_show_x(out, ((jl_uniontype_t*)v)->a, depth);
+            n += jl_printf(out, ", ");
+            v = ((jl_uniontype_t*)v)->b;
+        }
+        n += jl_static_show_x(out, v, depth);
+        n += jl_printf(out, "}");
+    }
+    else if (vt == jl_unionall_type) {
+        jl_unionall_t *ua = (jl_unionall_t*)v;
+        jl_tvar_t *var = ua->var;
+        n += jl_static_show_x(out, ua->body, depth);
+        n += jl_printf(out, " where ");
+        jl_value_t *lb = var->lb, *ub = var->ub;
+        if (lb != jl_bottom_type) {
+            if (jl_is_unionall(lb)) n += jl_printf(out, "(");
+            n += jl_static_show(out, lb);
+            if (jl_is_unionall(lb)) n += jl_printf(out, ")");
             n += jl_printf(out, "<:");
         }
-        n += jl_printf(out, "%s%s<:", (((jl_tvar_t*)v)->bound)?"#":"",
-                       jl_symbol_name(((jl_tvar_t*)v)->name));
-        n += jl_static_show(out, ((jl_tvar_t*)v)->ub);
+        n += jl_printf(out, "%s<:", jl_symbol_name(var->name));
+        if (jl_is_unionall(ub)) n += jl_printf(out, "(");
+        n += jl_static_show(out, ub);
+        if (jl_is_unionall(ub)) n += jl_printf(out, ")");
+    }
+    else if (vt == jl_tvar_type) {
+        n += jl_printf(out, "%s", jl_symbol_name(((jl_tvar_t*)v)->name));
```

Comment:
I wanted to avoid printing the bounds on every occurrence of a tvar. Do you think we should?